interp/interp.go

```diff
@@ -146,18 +153,148 @@ func (pc PiecewiseConstant) Predict(x float64) float64 {
 	if i < 0 {
 		return pc.ys[0]
 	}
-	// i < len(pci.xs)
+	// i < len(pc.xs)
 	if x == pc.xs[i] {
 		return pc.ys[i]
 	}
 	n := len(pc.xs)
 	if i == n-1 {
-		// x > pci.xs[i]
+		// x > pc.xs[i]
 		return pc.ys[n-1]
 	}
 	return pc.ys[i+1]
 }
 
+// PiecewiseCubic is a left-continuous, piecewise cubic
+// 1-dimensional interpolator.
+type PiecewiseCubic struct {
+	// Interpolated X values.
+	xs []float64
+
+	// Coefficients of interpolating cubic polynomials, with
+	// len(xs) - 1 rows and 4 columns. The interpolated value
+	// for xs[i] <= x < xs[i + 1] is defined as
+	//   sum_{k = 0}^3 coeffs.At(i, k) * (x - xs[i])^k
+	// To guarantee left-continuity, coeffs.At(i, 0) == ys[i].
+	coeffs *mat.Dense
+
+	// Last interpolated Y value, corresponding to xs[len(xs) - 1].
+	lastY float64
+}
+
+// Predict returns the interpolation value at x.
+func (pc PiecewiseCubic) Predict(x float64) float64 {
+	i := findSegment(pc.xs, x)
+	if i < 0 {
+		return pc.coeffs.At(0, 0)
+	}
+	m := len(pc.xs) - 1
+	// i < len(pc.xs)
+	if x == pc.xs[i] {
+		if i < m {
+			return pc.coeffs.At(i, 0)
+		}
+		return pc.lastY
+	}
+	if i == m {
+		// x > pc.xs[i]
+		return pc.lastY
+	}
+	dx := x - pc.xs[i]
+	a := pc.coeffs.RawRowView(i)
+	return ((a[3]*dx+a[2])*dx+a[1])*dx + a[0]
+}
+
+// FitWithDerivatives fits a piecewise cubic predictor to (X, Y, dY/dX) value
+// triples provided as three slices.
+// It returns an error if len(xs) < 2, elements of xs are not strictly
+// increasing, len(xs) != len(ys) or len(xs) != len(dydxs).
```

Comment:
Returning errors was requested by @btracey when reviewing the previous PR for interp/.

Reply:
I'll have a look at the conversation.

Reply:
The motivation was that if you wanted something like Least Squares, doing the matrix inversion can cause an error. It seemed like supporting Least Squares is a necessary condition for fitting. Maybe this isn't needed in 1-d though.

Reply:
Thanks. Yes, if there is an errorable case that can occur during the calculation but can't be determined prior to that, those would return an error. These cases can be known prior, so they should panic. Whether there are errorable inputs it something that you two are more likely to know.

Reply:
In this case I will revert to panicking on preventable errors, but leave the option to return an error from Fit. Does it work for everyone?

Reply:
That works for me. It would be nice if there were some examples of places where it's likely to be needed so we can make a clearer decision. At the moment, we're not in a release, so it's easy to change. After Augustish this will have been released and so it will be harder (though not impossible by any means) to remove.

Reply:
OK.

Reply:
This distinction between panicking on predictable error conditions and returning errors on unpredictable ones reminds me the distinction between unchecked and checked exceptions in Java, BTW.